core/drivers/clk/clk-stm32mp25.c

```diff
@@ -0,0 +1,3447 @@
+	uint32_t reg = 0;
+	uint32_t val = 0;
+
+	if (!id)
+		reg = RCC_FCALCOBS0CFGR;
+	else
+		reg = RCC_FCALCOBS1CFGR;
+
+	if (status)
+		val |= RCC_FCALCOBS0CFGR_CKOBSEN;
+
+	if (int_ext == OBS_EXT) {
+		val |= RCC_FCALCOBS0CFGR_CKOBSEXTSEL;
+		val |= SHIFT_U32(sel, RCC_FCALCOBS0CFGR_CKEXTSEL_SHIFT);
+	} else {
+		val |= SHIFT_U32(sel, RCC_FCALCOBS0CFGR_CKINTSEL_SHIFT);
+	}
+
+	if (inv)
+		val |= RCC_FCALCOBS0CFGR_CKOBSINV;
+
+	val |= SHIFT_U32(div, RCC_FCALCOBS0CFGR_CKOBSDIV_SHIFT);
+
+	io_write32(priv->base + reg, val);
+}
+
+static int stm32_clk_configure(struct clk_stm32_priv *priv, uint32_t val)
+{
+	uint32_t cmd_data = 0;
+	uint32_t cmd = 0;
+	int ret = 0;
+
+	if (val & CMD_ADDR_BIT) {
+		cmd_data = val & ~CMD_ADDR_BIT;
+
+		return stm32_clk_configure_by_addr_val(priv, cmd_data);
+	}
+
+	cmd = (val & CMD_MASK) >> CMD_SHIFT;
+	cmd_data = val & ~CMD_MASK;
+
+	switch (cmd) {
+	case CMD_DIV:
+		ret = stm32_clk_configure_div(priv, cmd_data);
+		break;
+
+	case CMD_MUX:
+		ret = stm32_clk_configure_mux(priv, cmd_data);
+		break;
+
+	case CMD_OBS:
+		stm32_clk_configure_obs(priv, cmd_data);
+		break;
+
+	default:
+		EMSG("cmd unknown ! : %#"PRIx32, val);
+		ret = -1;
+	}
+
+	return ret;
+}
+
+static int stm32_clk_bus_configure(struct clk_stm32_priv *priv)
+{
+	struct stm32_clk_platdata *pdata = priv->pdata;
+	uint32_t i = 0;
+
+	for (i = 0; i < pdata->nbusclk; i++) {
+		int ret = 0;
+
+		ret = stm32_clk_configure(priv, pdata->busclk[i]);
+		if (ret != 0)
+			return ret;
+	}
+
+	return 0;
+}
+
+static int stm32_clk_kernel_configure(struct clk_stm32_priv *priv)
+{
+	struct stm32_clk_platdata *pdata = priv->pdata;
+	uint32_t i = 0;
+
+	for (i = 0; i < pdata->nkernelclk; i++) {
+		int ret = 0;
+
+		ret = stm32_clk_configure(priv, pdata->kernelclk[i]);
+		if (ret != 0)
+			return ret;
+	}
+
+	return 0;
+}
+
+static void stm32mp2_init_clock_tree(struct clk_stm32_priv *priv,
+				     struct stm32_clk_platdata *pdata)
+{
+	stm32_clk_oscillators_enable(priv, pdata);
+
+	/* Come back to HSI for flexgen */
+	stm32mp2_clk_xbar_on_hsi(priv);
+
+	if (stm32_clk_pll_configure(priv))
+		panic("Cannot configure plls");
+
+	/* Wait LSE ready before to use it */
+	if (stm32_clk_oscillators_wait_lse_ready(priv, pdata))
+		panic("Timeout: to enable LSE");
+
+	if (stm32mp2_clk_flexgen_configure(priv))
+		panic("Cannot configure flexgen");
+
+	if (stm32_clk_bus_configure(priv))
+		panic("Cannot config bus clocks");
+
+	if (stm32_clk_kernel_configure(priv))
+		panic("Cannot configure kernel clocks");
+
+	/* Configure LSE css after RTC source configuration */
+	stm32_clk_oscillators_lse_set_css(priv, pdata);
+}
+
+static TEE_Result clk_stm32_osc_enable(struct clk *clk)
+{
+	return clk_stm32_gate_ready_ops.enable(clk);
+}
+
+static void clk_stm32_osc_disable(struct clk *clk)
+{
+	clk_stm32_gate_ready_ops.disable(clk);
+}
+
+static const struct clk_ops clk_stm32_osc_ops = {
+	.enable		= clk_stm32_osc_enable,
+	.disable	= clk_stm32_osc_disable,
```

Comment:
Could discard the inner indentation. 
Ditto at lines 1890/1893, 1931/1933, 2023/2025 etc...